### PR TITLE
feat: [CDS-68612]: support secrets expression for terraform remote var files

### PR DESCRIPTION
**Pull request by vlica-harness**
## Describe your changes
Still working on unit-tests for new V2 terraform steps.

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/48579)

### CODE REVIEW

**Comment by vlica-harness** [comment]
trigger smartchecks

**Comment by vlica-harness** [comment]
trigger uts

**Comment by vlica-harness** [comment]
trigger smartchecks

**Comment by vlica-harness** [comment]
trigger uts

**Comment by vlica-harness** [comment]
trigger uts

**Comment by vlica-harness** [comment]
trigger uts

**Review by tmacari-harness — CHANGES_REQUESTED** [review]
> Reviewed 3 of 15 files at r1, 13 of 25 files at r2, 10 of 16 files at r3, 2 of 5 files at r5, 7 of 7 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48579)* status: 33 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraform/TerraformPassThroughData.java` line 34 at r6](https://reviewable.io/reviews/harness/harness-core/48579#-NYwIH3cBHut4_MwjhZJ:-NYwIH3cBHut4_MwjhZK:b-bm6nxu) ([raw file](https://github.com/harness/harness-core/blob/7f6233c849d51af390cd8cc4df1a44ed412f9971/125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraform/TerraformPassThroughData.java#L34)):*
> ```Java
>   @Builder.Default Map<String, String> fetchedCommitIdsMap = new HashMap<>();
>   @Builder.Default Map<String, Map<String, String>> keyVersionMap = new HashMap<>();
>   @Builder.Default List<String> remoteFilesContent = new ArrayList<>();
> ```

maybe remoteVarFilesContent would be more clear... in future we may need to de the same for remote backend files
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraform/TerraformStepHelper.java` line 347 at r6](https://reviewable.io/reviews/harness/harness-core/48579#-NYwJom8BvrpxDBKsMfK:-NYwJom8BvrpxDBKsMfL:b-ilfytr) ([raw file](https://github.com/harness/harness-core/blob/7f6233c849d51af390cd8cc4df1a44ed412f9971/125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraform/TerraformStepHelper.java#L347)):*
> ```Java
>     GitFetchFilesConfigBuilder builder = GitFetchFilesConfig.builder();
> 
>     if (manifestType != null) {
> ```

Do we need that if?
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraform/TerraformStepHelper.java` line 1752 at r6](https://reviewable.io/reviews/harness/harness-core/48579#-NYwM1dH1IKHgUMBg7Ab:-NYwM1dH1IKHgUMBg7Ac:b-6cpayk) ([raw file](https://github.com/harness/harness-core/blob/7f6233c849d51af390cd8cc4df1a44ed412f9971/125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraform/TerraformStepHelper.java#L1752)):*
> ```Java
>       }
>     } else {
>       String errorMessage = "Error fetching remote terraform var-files";
> ```

Looks like we already check if we have s3 or git var files when we call that method. Not sure if we need that check here and return response with that error message.
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraform/steps/rolllback/TerraformRollbackStepV2.java` line 141 at r6](https://reviewable.io/reviews/harness/harness-core/48579#-NYwNXnZDfB8moFNeFZM:-NYwNXnZDfB8moFNeFZN:b-h4jmod) ([raw file](https://github.com/harness/harness-core/blob/7f6233c849d51af390cd8cc4df1a44ed412f9971/125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraform/steps/rolllback/TerraformRollbackStepV2.java#L141)):*
> ```Java
>       ExpressionEvaluatorUtils.updateExpressions(
>           rollbackConfig, new CDExpressionResolveFunctor(engineExpressionService, ambiance));
>       // TODO:  log rollback message
> ```

Let's remove that TODO, if we k now what should be done ... we can do it now.

**Comment by vlica-harness** [comment]
*[`125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraform/TerraformStepHelper.java` line 347 at r6](https://reviewable.io/reviews/harness/harness-core/48579#-NYwJom8BvrpxDBKsMfK:-NYxTu5B5nhg-ZRbvNID:b9l50bo) ([raw file](https://github.com/harness/harness-core/blob/7f6233c849d51af390cd8cc4df1a44ed412f9971/125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraform/TerraformStepHelper.java#L347)):*
<details><summary><i>Previously, tmacari-harness (Tudor Macari) wrote…</i></summary><blockquote>

Do we need that if?
</blockquote></details>

yes,  this parameter  is sent only from V2 steps. Didn't want to change the logic used in original steps. Even it's smth minor.

**Comment by vlica-harness** [comment]
*[`125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraform/TerraformStepHelper.java` line 1752 at r6](https://reviewable.io/reviews/harness/harness-core/48579#-NYwM1dH1IKHgUMBg7Ab:-NYxUHvB8Yp-cIYQcl8y:bny3n85) ([raw file](https://github.com/harness/harness-core/blob/7f6233c849d51af390cd8cc4df1a44ed412f9971/125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraform/TerraformStepHelper.java#L1752)):*
<details><summary><i>Previously, tmacari-harness (Tudor Macari) wrote…</i></summary><blockquote>

Looks like we already check if we have s3 or git var files when we call that method. Not sure if we need that check here and return response with that error message.
</blockquote></details>

yes, this logic is to ensure no other types are gonna be handled in this method, it's just a default response.

**Comment by vlica-harness** [comment]
*[`125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraform/steps/rolllback/TerraformRollbackStepV2.java` line 141 at r6](https://reviewable.io/reviews/harness/harness-core/48579#-NYwNXnZDfB8moFNeFZM:-NYxUtsx5XZyktd42Y1D:b-hfpv0l) ([raw file](https://github.com/harness/harness-core/blob/7f6233c849d51af390cd8cc4df1a44ed412f9971/125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraform/steps/rolllback/TerraformRollbackStepV2.java#L141)):*
<details><summary><i>Previously, tmacari-harness (Tudor Macari) wrote…</i></summary><blockquote>

Let's remove that TODO, if we k now what should be done ... we can do it now.
</blockquote></details>

removed

**Review by vlica-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48579)* status: 35 file(s) to review. 1 change requests. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraform/TerraformPassThroughData.java` line 34 at r6](https://reviewable.io/reviews/harness/harness-core/48579#-NYwIH3cBHut4_MwjhZJ:-NYxTp7k2HON2TuxQBcg:bejeyqs) ([raw file](https://github.com/harness/harness-core/blob/7f6233c849d51af390cd8cc4df1a44ed412f9971/125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraform/TerraformPassThroughData.java#L34)):*
<details><summary><i>Previously, tmacari-harness (Tudor Macari) wrote…</i></summary><blockquote>

maybe remoteVarFilesContent would be more clear... in future we may need to de the same for remote backend files
</blockquote></details>

makes sense, done

**Comment by vlica-harness** [comment]
trigger uts

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/aws/S3FetchFilesTaskNG.java [hunk 1]
```diff
@@ -137,12 +161,22 @@ private List<S3FileDetailResponse> getContentFromFromS3Bucket(
         executionLogCallback.saveExecutionLog(
             format("%nSuccessfully fetched %s file from s3 bucket: %s", fileKey, bucketName));
       } catch (Exception e) {
-        String errorMsg = format("Failed to fetch %s from s3 bucket: %s : %s", fileKey, bucketName, e.getMessage());
+        String exceptionMessage = "";
+        if (e instanceof WingsException) {
+          WingsException wingsException = (WingsException) e;
+          if (wingsException.getParams() != null && wingsException.getParams().get("message") != null) {
+            exceptionMessage = wingsException.getParams().get("message").toString();
+          }
```
Comment (yogesh-chauhan) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/aws/S3FetchFilesTaskNG.java:169]: not sure but does ExceptionUtils.getMessage() also do something similar ?

**Review by yogesh-chauhan — APPROVED** [review]

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/aws/S3FetchFilesTaskNG.java (hunk 1 above)
Comment (vlica-harness) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/aws/S3FetchFilesTaskNG.java:169]: The logic in ExceptionUtils is a bit complex, it's getting the message, but before that it's doing other alters with the params, will leave as it is atm.

**Comment by vlica-harness** [comment]
trigger uts

**Review by tmacari-harness — APPROVED** [review]
> Reviewed 1 of 7 files at r7, 15 of 18 files at r8, 3 of 3 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48579)* status: 36 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 3 of 15 files at r1, 8 of 25 files at r2, 7 of 16 files at r3, 1 of 5 files at r5, 1 of 7 files at r7, 15 of 18 files at r8, 3 of 3 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48579)* status: :shipit: complete! ✓

**Review by yogesh-chauhan — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48579)* status: :shipit: complete! ✓

**Comment by vlica-harness** [comment]
trigger review

**Comment by vlica-harness** [comment]
trigger review

**Pull request merged**